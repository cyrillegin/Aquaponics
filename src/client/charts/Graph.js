/* eslint-disable id-length */
import React, { useEffect, useState, useRef } from 'react';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import * as d3 from 'd3';
import {
  searchToObject,
  objectToString,
  windowEmitter,
  addOrUpdateHash,
} from '../utilities/Window';

const getReadings = (sensor, setReadings) => {
  const search = searchToObject();
  const kwargs = {
    sensorId: sensor.id,
  };
  if (search.start) {
    kwargs.start = search.start;
  }
  if (search.end) {
    kwargs.end = search.end;
  }

  fetch(`/api/reading?${objectToString(kwargs)}`)
    .then(res => res.json())
    .then(newReadings => {
      setReadings(
        newReadings
          .map(reading => ({ ...reading, timestamp: new Date(reading.timestamp) }))
          // Sorting is redundant as it takes place in the db,
          // leaving here though just in case :)
          .sort((a, b) => a.timestamp - b.timestamp),
      );
    });
};

const Graph = ({ className, station, sensor, renderTrigger }) => {
  const graphElement = useRef(null);
  const [readings, setReadings] = useState([]);

  useEffect(() => {
    getReadings(sensor, setReadings);
  }, [sensor, renderTrigger]);

  if (graphElement.current) {
    const margin = { top: 50, right: 50, bottom: 50, left: 50 };
    const width = graphElement.current.clientWidth - margin.left - margin.right;
    const height = graphElement.current.clientHeight - margin.top - margin.bottom;

    const xScale = d3
      .scaleTime()
      .range([0, width])
      .domain(d3.extent(readings, d => d.timestamp));

    const yScale = d3
      .scaleLinear()
      .range([height, 0])
      .domain([d3.min(readings, d => d.value), d3.max(readings, d => d.value)]);

    const valueline = d3
      .line()
      .x(d => xScale(d.timestamp))
      .y(d => yScale(d.value));

    const svg = d3
      .select(graphElement.current)
      .html('')
      .append('svg')
      .attr('width', width + margin.left + margin.right)
      .attr('height', height + margin.top + margin.bottom)
      .append('g')
      .attr('transform', `translate(${margin.left},${margin.top})`);

    // Title
    svg.append('text').text(`${station.name} - ${sensor.name}`).attr('class', 'graph-title');

    // Bottom axis
    svg
      .append('g')
      .attr('class', 'x-axis')
      .attr('transform', `translate(0,${height})`)
      .call(d3.axisBottom(xScale).tickFormat(d3.timeFormat('%Y-%m-%d')));

    // Left axis
    svg
      .append('line')
      .attr('x1', 0.5)
      .attr('y1', 0)
      .attr('x2', 0.5)
      .attr('y2', height)
      .attr('stroke', 'currentcolor');

    // Left axis ticks
    svg
      .append('g')
      .attr('class', 'y-axis')
      .call(d3.axisLeft(yScale).tickSize(width - margin.left - margin.right))
      .call(g => g.select('.domain').remove())
      .call(g => g.selectAll('.tick line').attr('x2', width).attr('class', 'tick-line'))
      .call(g => g.selectAll('.tick text').attr('x', -4));

    // Right axis
    svg
      .append('line')
      .attr('x1', width + 0.5)
      .attr('y1', 0)
      .attr('x2', width + 0.5)
      .attr('y2', height)
      .attr('stroke', 'currentcolor');

    // Top axis
    svg
      .append('line')
      .attr('x1', 0)
      .attr('y1', 0.5)
      .attr('x2', width + 0.5)
      .attr('y2', 0.5)
      .attr('stroke', 'currentcolor');

    // Line graph
    svg.append('path').data([readings]).attr('class', 'line').attr('d', valueline);

    if (readings.length === 0) {
      // No readings text
      svg
        .append('text')
        .text('No readings exist in this range :(')
        .attr('transform', `translate(${width / 2 - 80},${height / 2})`);

      return (
        <div ref={graphElement} className={className}>
          <div id="my-svg"> </div>
        </div>
      );
    }
    const tooltip = svg
      .append('g')
      .attr('class', 'tooltip')
      .attr('transform', `translate(${margin.left}, 0)`);

    // Tooltip line
    tooltip.append('line').attr('x1', 0).attr('y1', 0).attr('x2', 0).attr('y2', height);

    // Tooltip circle
    tooltip.append('circle').attr('cx', 0).attr('cy', 0).attr('r', 5);

    // Tooltip text
    tooltip.append('text');

    // Handle tooltip movement
    // non arrow function required here for 'this' scope
    d3.select('svg').on('mousemove', function () {
      const x0 = xScale.invert(d3.mouse(this)[0] - margin.left);
      let i = d3.bisector(d => d.timestamp).right(readings, x0, 1);
      // Prevents out of bounds exception
      if (i > readings.length - 1) {
        i = readings.length - 1;
      }

      tooltip
        .select('circle')
        .attr(
          'transform',
          `translate(${xScale(readings[i].timestamp) - margin.right}, ${yScale(
            readings[i].value,
          )})`,
        );

      tooltip
        .select('line')
        .attr('transform', `translate(${xScale(readings[i].timestamp) - margin.right}, 0)`);

      tooltip
        .select('text')
        .text(readings[i].value.toFixed(2))
        .attr(
          'transform',
          `translate(${xScale(readings[i].timestamp) - margin.right + 10}, ${yScale(
            readings[i].value,
          )})`,
        );
    });

    // Drag functions
    const selectionBox = svg
      .append('rect')
      .attr('fill', 'none')
      .attr('opacity', 0.5)
      .attr('x', 0)
      .attr('y', 0)
      .attr('width', 14)
      .attr('height', height);

    let dragStart = 0;
    let dragStartPos = 0;
    let dragEnd = 0;
    const drag = d3
      .drag()
      .on('drag', function (d, i) {
        const x0 = xScale.invert(d3.mouse(this)[0]);
        let index = d3.bisector(datum => datum.timestamp).right(readings, x0, 1);
        // Prevents out of bounds exception
        if (i > readings.length - 1) {
          index = readings.length - 1;
        }

        if (xScale(readings[index].timestamp) - margin.left > dragStartPos + margin.right) {
          selectionBox.attr(
            'width',
            xScale(readings[index].timestamp) - dragStartPos - margin.left,
          );
        } else {
          selectionBox.attr(
            'width',
            dragStartPos + margin.left - xScale(readings[index].timestamp),
          );
          selectionBox.attr(
            'transform',
            `translate(${xScale(readings[index].timestamp) - margin.left},0)`,
          );
        }
      })
      .on('end', function (d, i) {
        dragEnd = d3.mouse(this)[0] - margin.right;
        if (Math.abs(dragStart - dragEnd) < 10) return;

        const x0 = xScale.invert(dragStart);
        const x1 = xScale.invert(dragEnd);

        if (x1 > x0) {
          addOrUpdateHash({ start: x0.toISOString(), end: x1.toISOString() });
        } else {
          addOrUpdateHash({ start: x1.toISOString(), end: x0.toISOString() });
        }
      });

    d3.select('svg')
      .on('mousedown', function () {
        selectionBox.attr('fill', '#b7ff64');
        dragStart = d3.mouse(this)[0] - margin.right;

        let index = d3
          .bisector(datum => datum.timestamp)
          .right(readings, xScale.invert(dragStart), 1);
        // Prevents out of bounds exception
        if (index > readings.length - 1) {
          index = readings.length - 1;
        }
        selectionBox.attr('transform', `translate(${xScale(readings[index].timestamp)},0)`);
        dragStartPos = xScale(readings[index].timestamp);
      })
      .call(drag);
  }

  return (
    <div ref={graphElement} className={className}>
      <div id="my-svg"> </div>
    </div>
  );
};

Graph.propTypes = {
  className: PropTypes.string.isRequired,
  station: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  sensor: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  /* This is used for the dashboard component to cause a
   * re-fetching of readings. Possible reasons are:
   * - new sensor
   * - new start time
   * - new end time
   */

  renderTrigger: PropTypes.instanceOf(Date).isRequired,
};

const styledGraph = styled(Graph)`
  width: 100%;
  height: 600px;
  background: white;

  .graph-title {
    transform: translate(0px, -8px);
  }

  .line {
    fill: none;
    stroke: #ffab00;
    stroke-width: 1.5px;
  }

  .tick-line {
    color: grey;
  }

  .tooltip {
    line {
      stroke: #3f94ff;
    }

    circle {
      stroke: #3f94ff;
      fill: none;
      stroke-width: 1;
    }
    text {
      stroke: #3f94ff;
    }
  }
`;

export default styledGraph;
