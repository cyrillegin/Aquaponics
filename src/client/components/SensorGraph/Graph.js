import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {withStyles} from '@material-ui/core/styles';
import queryString from 'query-string';
import './graph-styles.scss';

const d3 = {
  ...require('d3-selection'), // eslint-disable-line
  ...require('d3-scale'), // eslint-disable-line
  ...require('d3-axis'), // eslint-disable-line
  ...require('d3-format'), // eslint-disable-line
  ...require('d3-shape'), // eslint-disable-line
  ...require('d3-array') // eslint-disable-line
};

const styles = theme => ({
  root: {},
});

export class Graph extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    sensor: PropTypes.object,
    readings: PropTypes.arrayOf(PropTypes.shape({
      timestamp: PropTypes.number.isRequired,
      value: PropTypes.number.isRequired,
    })).isRequired,
  }

  static defaultProps = {
    sensor: null,
  }

  drawGraph() {
    const data = this.props.readings;
    // Initialization.
    const container = document.querySelector('#graph-container');
    container.innerHTML = '';

    let width = container.clientWidth;

    let height = 400;
    const margin = {
      top: 10,
      right: 20,
      bottom: 10,
      left: this.props.sensor.units ? this.props.sensor.units.length * 10 : 60,
    };
    width = width - margin.left - margin.right;
    height = height - margin.top - margin.bottom;
    let i = 0;
    let newText = '';

    // // Apply calibration data.
    const coef = {
      x: 1,
      y: 0,
    };

    const dataObject = [];
    for (i = data.length - 1; i > 0; i --) {
      dataObject.push({
        created: data[i].timestamp,
        value: data[i].value * coef.x + coef.y,
      });
    }
    if (dataObject.length === 0) {
      container.innerHTML = 'There arnt enough readings for this sensor to display anything.';
      return;
    }
    let start = queryString.parse(location.search).start;
    let end = queryString.parse(location.search).end;
    if (end === undefined) {
      end = Date.now();
    } else {
      end = new Date(parseInt(end));
    }
    console.log(start, end);
    if (start === undefined) {
      start = Date.now() - 24 * 60 * 60 * 1000;
    } else {
      start = new Date(parseInt(start));
    }
    let min = dataObject[0].value;
    let max = dataObject[0].value;

    // Set min and max values
    for (i = 0; i < dataObject.length; i ++) {
      if (min > dataObject[i].value) {
        min = dataObject[i].value;
      }
      if (max < dataObject[i].value) {
        max = dataObject[i].value;
      }
    }

    // Create the svg.
    const newChart = d3.select('#graph-container')
      .append('svg')
      .attr('width', width + margin.left + margin.right)
      .attr('height', height + margin.top + margin.bottom)
      .append('g')
      .attr('transform', `translate(${margin.left},${margin.bottom})`)
      .classed('svg-content-responsive', true);

    // Scale
    const xScale = d3.scaleTime()
      .domain([new Date(start), new Date(end)])
      .rangeRound([0, width]);

    const yScale = d3.scaleLinear()
      .domain([min, max])
      .rangeRound([height - margin.bottom, margin.top]);

    // Formats text for units display
    const getFormattedText = (d) => {
      const f = d3.format('.1f');
      const units = this.props.sensor.units;

      return f(d) + units;
    };

    // TOOL-TIPS
    // Tooltip container
    const tooltip = newChart.append('g')
      .style('display', 'none');

    // Tooltip helper
    const bisectDate = d3.bisector((d) => d.created).left;

    // Y-axis line for tooltip
    const yLine = tooltip.append('g')
      .append('line')
      .attr('class', 'tooltip-line')
      .style('stroke', 'blue')
      .style('stroke-dasharray', '3,3')
      .style('opacity', 0.5)
      .attr('y1', margin.bottom)
      .attr('y2', height);

    // Date text
    const timeText = tooltip.append('text')
      .attr('x', 0)
      .attr('y', margin.top - 5)
      .attr('width', 100)
      .attr('height', 100 * 0.4)
      .attr('fill', 'black');

    // Drag behaivors for the selection box.
    // const dragStart = 0, dragStartPos = 0, dragEnd = 0;
    // const drag = d3.drag()
    //     .on('drag', function(d,i) {
    //         const x0 = xScale.invert(d3.mouse(this)[0]);
    //         i = bisectDate(data.readings, x0, 1);
    //         const d0 = data.readings[i - 1],
    //             d1 = data.readings[i];
    //         d = x0 - d0.created > d1.created - x0 ? d1 : d0;

    //         if(xScale(d.created) > dragStartPos){
    //             selectionBox.attr('width', (xScale(d.created) - dragStartPos));
    //         } else {
    //             selectionBox.attr('width', ( dragStartPos - xScale(d.created)));
    //             selectionBox.attr('transform', 'translate(' + xScale(d.created) + ',0)' );
    //         }
    //     })
    //     .on('end', function(d,i){
    //         dragEnd = d3.mouse(this)[0];
    //         if(Math.abs(dragStart - dragEnd) < 10) return;

    //         const x0 = xScale.invert(dragStart), x1 = xScale.invert(dragEnd);

    // scope.$apply(function(){
    //     if(x1 > x0){
    //         $location.search('start_date', x0.getTime());
    //         $location.search('end_date',x1.getTime());
    //     } else {
    //         $location.search('start_date', x1.getTime());
    //         $location.search('end_date',x0.getTime());
    //     }
    // });
    // });
    // Update loop for tooltips.
    const circleElements = [];
    const textElements = [];

    function mousemove() {
      const x0 = xScale.invert(d3.mouse(this)[0]); // jshint ignore:line
      const cross = bisectDate(dataObject, x0, 1);
      const d0 = dataObject[cross - 1];
      const d1 = dataObject[cross];
      if (d1 === undefined) {
        return;
      }
      const d = x0 - d0.created > d1.created - x0 ? d1 : d0;

      circleElements[0].attr('transform', `translate(${xScale(d.created)},${yScale(d.value)})`);
      yLine.attr('transform', `translate(${xScale(d.created)},0)`);
      timeText.text(new Date(d.created));

      textElements[0]
        .text(getFormattedText(d.value))
        .attr('transform', `translate(${(xScale(d.created) + 10)},${(yScale(d.value) - 10)})`);
    }

    // Selection box
    const selectionBox = newChart.append('rect')
      .attr('fill', 'none')
      .attr('opacity', 0.5)
      .attr('x', 0)
      .attr('y', margin.top)
      .attr('width', 14)
      .attr('height', height - margin.top);

    // Hit area for selection box
    newChart.append('rect')
      .attr('width', width)
      .attr('height', height)
      .style('fill', 'none')
      .style('pointer-events', 'all')
      .on('mouseover', () => {
        tooltip.style('display', null);
      })
      .on('mouseout', () => {
        tooltip.style('display', 'none');
      })
      .on('mousemove', mousemove)
      .on('mousedown', () => {
        selectionBox.attr('fill', '#b7ff64');
        // dragStart = d3.mouse(this)[0];

        const x0 = xScale.invert(d3.mouse(this)[0]);
        const cross = bisectDate(dataObject.readings, x0, 1);
        const d0 = dataObject.readings[cross - 1];
        const d1 = dataObject.readings[cross];
        const d = x0 - d0.created > d1.created - x0 ? d1 : d0;
        selectionBox.attr('transform', `translate(${xScale(d.created)},0)`);
        // dragStartPos = xScale(d.created);
      });
    // .call(drag);

    // Y Axis
    function getTic() {
      const Ticks = [];
      const ratio = (max - min) / 6;
      for (let q = 0; q < 7; q ++) {
        Ticks.push(min + (ratio * q));
      }
      return Ticks;
    }

    const yAxis = d3.axisLeft(yScale)
      .tickSizeInner(- width)
      .tickSizeOuter(0)
      .tickValues(getTic())
      .tickFormat((d) => {
        const f = d3.format('.1f');
        return `${f(d)} ${this.props.sensor.units}`;
      });

    newChart.append('g')
      .attr('class', 'chart-axis-shape')
      .call(yAxis);

    // X Axis
    const xAxis = d3.axisBottom(xScale)
      .tickSizeInner(- height + margin.bottom + margin.top)
      .tickSizeOuter(0)
      .tickPadding(10)
      .ticks(10);

    newChart.append('g')
      .attr('class', 'chart-axis-shape')
      .attr('transform', `translate(0,${height - margin.bottom})`)
      .call(xAxis);

    // Top border
    newChart.append('g')
      .append('line')
      .attr('class', 'chart-axis-shape')
      .attr('x1', 0)
      .attr('x2', width)
      .attr('y1', margin.bottom)
      .attr('y2', margin.bottom);

    // Right border
    newChart.append('g')
      .append('line')
      .attr('class', 'chart-axis-shape')
      .attr('x1', width)
      .attr('x2', width)
      .attr('y1', margin.bottom)
      .attr('y2', height - margin.bottom);

    // Graph lines
    const lineFunction = d3.line()
      // .defined(() => {
      //     // TODO: add linebreak behaivor
      //     return true;
      // })
      .x((d) => xScale(d.created))
      .y((d) => yScale(d.value));

    newChart.append('path')
      .attr('d', lineFunction(dataObject))
      .attr('stroke', 'rgba(45, 100, 255, 0.87)')
      .attr('stroke-width', 2)
      .attr('fill', 'none');

    // // Tooltip circle
    const newCircle = tooltip.append('circle')
      .style('fill', 'none')
      .style('stroke', 'blue')
      .attr('r', 4);
    circleElements.push(newCircle);

    // Tooltip text
    newText = tooltip.append('text')
      .attr('width', 100 * 2)
      .attr('height', 100 * 0.4)
      .attr('fill', 'black');
    textElements.push(newText);
  } // End D3

  componentDidMount() {
    const data = [];
    for (let i = 0; i < 24 * 60; i++) {
      data.push({
        timestamp: Date.now() - (i * 60 * 1000),
        value: Math.sin(i / 100),
      });
    }
    this.drawGraph(data);
  }


  render() {
    return (
      <div className={this.props.classes.root}>
        <div id="graph-container" />
      </div>
    );
  }
}

export default withStyles(styles)(Graph);
