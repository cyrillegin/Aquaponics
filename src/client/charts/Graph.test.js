import { mount } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';
import Graph from './Graph';

describe('dashboard', () => {
  it('should render a snap shot', async () => {
    global.fetch = jest.fn(
      () =>
        new Promise(res =>
          res({ json: () => new Promise(inner => inner([{ timestamp: new Date(), value: 1 }])) }),
        ),
    );

    let wrapper;
    await act(async () => {
      wrapper = await mount(
        <Graph name="test" sensor={{ id: 1, name: 'test' }} renderTrigger={new Date()} />,
      );
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
