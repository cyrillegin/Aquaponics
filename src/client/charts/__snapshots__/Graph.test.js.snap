// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard should render a snap shot 1`] = `"<div class=\\"css-15c4fvq\\"><div class=\\"graph-container\\"><svg width=\\"0\\" height=\\"0\\"><g transform=\\"translate(50,50)\\"><g class=\\"x-axis\\" transform=\\"translate(0,-100)\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"middle\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M0.5,6V0.5H-99.5V6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(-49.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">2021-01-02</text></g></g><line x1=\\"0.5\\" y1=\\"0\\" x2=\\"0.5\\" y2=\\"-100\\" stroke=\\"currentcolor\\"></line><g class=\\"y-axis\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"end\\"><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,-49.5)\\"><line stroke=\\"currentColor\\" x2=\\"-100\\" class=\\"tick-line\\"></line><text fill=\\"currentColor\\" x=\\"-4\\" dy=\\"0.32em\\">1</text></g></g><line x1=\\"-99.5\\" y1=\\"0\\" x2=\\"-99.5\\" y2=\\"-100\\" stroke=\\"currentcolor\\"></line><line x1=\\"0\\" y1=\\"0.5\\" x2=\\"-99.5\\" y2=\\"0.5\\" stroke=\\"currentcolor\\"></line><path class=\\"line\\" d=\\"M-50,-50Z\\"></path><text class=\\"graph-title\\">test - test</text><g class=\\"tooltip\\" transform=\\"translate(50, 0)\\" display=\\"none\\"><line x1=\\"0\\" y1=\\"0\\" x2=\\"0\\" y2=\\"-100\\"></line><circle cx=\\"0\\" cy=\\"0\\" r=\\"5\\"></circle><text></text></g><rect fill=\\"none\\" opacity=\\"0.5\\" x=\\"0\\" y=\\"0\\" width=\\"14\\" height=\\"-100\\"></rect></g></svg></div></div>"`;
